src/Microsoft.SqlTools.Authentication/Authenticator.cs

```diff
@@ -0,0 +1,242 @@
+                                break;
+                            }
+                        }
+                        while (accounts.MoveNext());
+                    }
+                    else
+                    {
+                        account = accounts.Current;
+                    }
+                }
+
+                if (null != account)
+                {
+                    try
+                    {
+                        // Fetch token silently
+                        result = await publicClientApplication.AcquireTokenSilent(@params.Scopes, account)
+                            .ExecuteAsync(cancellationToken: cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                    catch (Exception e) when (e is MsalUiRequiredException || e is MsalServiceException)
+                    {
+                        result = await AcquireTokenInteractiveAsync(publicClientApplication,
+                            @params.Scopes, @params.ConnectionId, username,
+                            @params.AuthenticationMethod, cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    // If no existing 'account' is found, we request user to sign in interactively.
+                    result = await AcquireTokenInteractiveAsync(publicClientApplication,
+                        @params.Scopes, @params.ConnectionId, username,
+                        @params.AuthenticationMethod, cancellationToken)
+                        .ConfigureAwait(false);
+                }
+            }
+            else
+            {
+                throw new Exception($"Authentication Method ${@params.AuthenticationMethod} is not supported.");
+            }
+
+            return result != null ? new AccessToken(result.AccessToken, result.ExpiresOn) : null;
+        }
+
+        #endregion
+
+        #region Private methods
+        private IPublicClientApplication GetPublicClientAppInstance(string authority, string audience)
+        {
+            string authorityUrl = authority + '/' + audience;
+            if (!s_pcaMap.TryGetValue(authorityUrl, out IPublicClientApplication? clientApplicationInstance))
+            {
+                clientApplicationInstance = CreatePublicClientAppInstance(authority, audience);
+                s_pcaMap.TryAdd(authorityUrl, clientApplicationInstance);
+            }
+            return clientApplicationInstance;
+        }
+
+        private IPublicClientApplication CreatePublicClientAppInstance(string authority, string audience) =>
+            PublicClientApplicationBuilder.Create(ApplicationClientId)
+                .WithAuthority(authority, audience)
+                .WithClientName(ApplicationName)
+                .WithRedirectUri(RedirectUri)
+                .WithClientId(ApplicationClientId)
+                .WithLogging((logLevel, message, pii) =>
```

Comment:
I thought we had a setting to turn on PII logging, IIRC.  Do we no longer want to support that or am I misremembering what we currently support?

Reply:
I couldn't find PII logging available in STS, we have the setting in Azure Core extension, I also thought it's generally not enabled when requesting logs from end users, so I disabled it.

We'll have to transfer setting here in order to enable it. Let me know if want me to implement it.

Reply:
thanks the details!  It's up to you whether it's worth implementing all the way through to here, but I do know that lack of logging has made several AAD related investigations very difficult.  And we added the PII logging to help with that.  It may be a good idea to anticipate getting issues where it is hard to repro the failure and the more detailed logs we can collect the better.